File: packages/api-client/src/dto/ticket.ts
/**
 * Wire-format DTOs for the Tickets domain.
 * Field names are snake_case on purpose — they mirror the JSON the API sends
 * (apps/api tickets module + docs/04-api/API_SPECIFICATION.md), not internal TS
 * naming. Status values are lowercase to match the backend TicketStatus enum.
 */
export type TicketStatus = 'available' | 'reserved' | 'paid' | 'winner' | 'cancelled';

export interface TicketDto {
  id: string;
  raffle_id: string;
  participant_id: string | null;
  collaborator_id: string | null;
  number: number;
  status: TicketStatus;
  reserved_at: string | null;
  expires_at: string | null;
  sold_at: string | null;
  winner_at: string | null;
  created_at: string;
  updated_at: string;
}

export interface ListTicketsQuery {
  [key: string]: string | number | boolean | undefined | null;
  page?: number;
  page_size?: number;
  raffle_id?: string | undefined;
  status?: TicketStatus | undefined;
}

/** Admin single-ticket reservation body; both fields optional. */
export interface ReserveTicketRequest {
  participant_id?: string;
  collaborator_id?: string;
}

File: apps/api/app/modules/tickets/schemas/ticket.py
import uuid
from datetime import datetime

from pydantic import BaseModel, ConfigDict

from app.modules.tickets.models import Ticket, TicketStatus


class ReserveTicketRequest(BaseModel):
    """Admin reservation body. ``participant_id`` and ``collaborator_id`` are both
    optional — a ticket may be reserved with or without a participant on record
    and with or without a collaborator (seller) credited."""

    model_config = ConfigDict(extra="forbid")

    participant_id: uuid.UUID | None = None
    collaborator_id: uuid.UUID | None = None


class SetCollaboratorRequest(BaseModel):
    """Body for crediting/clearing the collaborator (seller) of a reserved
    ticket. ``collaborator_id = null`` clears it."""

    model_config = ConfigDict(extra="forbid")

    collaborator_id: uuid.UUID | None = None


class AssignParticipantRequest(BaseModel):
    """Body for assigning/changing the participant of a ticket."""

    model_config = ConfigDict(extra="forbid")

    participant_id: uuid.UUID


class TicketRead(BaseModel):
    id: uuid.UUID
    raffle_id: uuid.UUID
    participant_id: uuid.UUID | None
    collaborator_id: uuid.UUID | None
    number: int
    status: TicketStatus
    reserved_at: datetime | None
    expires_at: datetime | None
    sold_at: datetime | None
    winner_at: datetime | None
    created_at: datetime
    updated_at: datetime

    @classmethod
    def from_entity(cls, ticket: Ticket) -> "TicketRead":
        return cls.model_validate(ticket, from_attributes=True)
